resolver/resolver.go

```diff
@@ -457,7 +457,12 @@ func isUDP(w dns.ResponseWriter) bool {
 // truncate removes answers until the given dns.Msg fits the permitted
 // length of the given transmission channel and sets the TC bit.
 // See https://tools.ietf.org/html/rfc1035#section-4.2.1
-func truncate(m *dns.Msg, udp bool) *dns.Msg {
+// If `setTruncateBit` is true, the Truncate bit will be set if the message
+// has been truncated already or gets truncated here.
+// If `setTruncateBit` is false, the message will have its Truncate bit
+// cleared if it is already set, and won't set it even if it does get
+// truncated.
+func truncate(m *dns.Msg, udp bool, setTruncateBit bool) *dns.Msg {
```

Comment:
Regarding the pointer - this is "c-style" code intended to be fast. It does not allocate new objects. If we were to return a new object instead of mutating the old one this function (and consequently mesos-dns as a whole) would be slower.

The reason we return the pointer to the input is to improve readability at the site where this call is made:

```
if err := w.WriteMsg(truncate(m, isUDP(w), setTruncateBit)); err != nil {
...
```
vs.
```
m = truncate(m, isUDP(w), setTruncateBit)
if err := w.WriteMsg(m); err != nil {
...
```

I'm easy either way - but this is the original author's reasoning behind this signature.

Regarding mutation and bugs in tests, see my comments lower down.